dev/src/bulk-writer.ts

```diff
@@ -740,27 +914,69 @@ export class BulkWriter {
    *
    * @private
    */
-  private sendBatch(batch: BulkCommitBatch): void {
+  private sendBatch(
+    batch: BulkCommitBatch,
+    isRetryBatch = false
+  ): Promise<void> {
     const success = this.rateLimiter.tryMakeRequest(batch.opCount);
     assert(success, 'Batch should be under rate limit to be sent.');
-    batch.bulkCommit().then(() => {
+    return batch.bulkCommit().then(() => {
       // Remove the batch from the BatchQueue after it has been processed.
-      const batchIndex = this.batchQueue.indexOf(batch);
+      const batchQueue = isRetryBatch
+        ? this._retryBatchQueue
+        : this._batchQueue;
+      const batchIndex = batchQueue.indexOf(batch);
       assert(batchIndex !== -1, 'The batch should be in the BatchQueue');
-      this.batchQueue.splice(batchIndex, 1);
+      batchQueue.splice(batchIndex, 1);
 
       this.sendReadyBatches();
     });
   }
 
+  private wrapRetry(
+    promise: Promise<WriteResult>,
+    documentRef: firestore.DocumentReference,
+    retryCount: number,
+    operationType: 'create' | 'set' | 'update' | 'delete',
+    retryFn: () => Promise<WriteResult>
+  ): Promise<WriteResult> {
+    // eslint-disable-next-line prefer-const
+    let voidOperationPromise: Promise<void>;
+    const operation = promise
+      .then(res => {
+        this._successFn(documentRef, res);
+        this._pendingOps.delete(voidOperationPromise);
+        return res;
+      })
+      .catch(error => {
+        const bulkWriterError = new BulkWriterError(
+          error.code,
+          error.message,
+          documentRef,
+          operationType,
+          retryCount
+        );
+        const shouldRetry = this._errorFn(bulkWriterError);
+        this._pendingOps.delete(voidOperationPromise);
+        if (shouldRetry) {
+          return retryFn();
+        } else {
+          throw bulkWriterError;
+        }
+      });
```

Comment:
Nit: This might be slightly prettier if we used try/catch with `async/await` here.